### PR TITLE
Settings panel refactor

### PR DESCRIPTION
**Pull request by haonliu**
#### Description of changes
Settings panel refactor
Move all and each setting to it's own component.
SettingsPanel only responsibility is to render those setting components; it has no knowledge about specifics of each setting (like the github details from before this refactor).

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Added relevant unit test for your changes. (`npm run test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] Ran precheckin (`npm run precheckin`)
- [ ] Added screenshots/GIFs for UI changes.

### CODE REVIEW

File: src/DetailsView/components/settings-panel/settings-panel.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (waabid) [inline src/DetailsView/components/settings-panel/settings-panel.tsx]: Name seems redundant if it's under 'settings-panel' folder. Maybe folder name should be 'settings'?

File: src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                className="issue-setting"
+                label="Enter desired GitHub repo link:"
+                onChange={onGitHubRepositoryChange}
+                value={getBugServiceProperty('gitHub', 'repository')}
+                placeholder="https://github.com/owner/repo/issues"
+            />
+        );
+    };
+
+    const onGitHubRepositoryChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
```
Comment (waabid) [inline src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx]: Are these methods not complaining of missing return type?

File: src/tests/unit/tests/DetailsView/components/settings-panel/settings-panel.test.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    type RenderTestCase = {
+        isPanelOpen: boolean;
+        enableTelemetry: boolean;
+        enableHighContrast: boolean;
+        bugService: string;
+        bugServicePropertiesMap: BugServicePropertiesMap;
+    };
+
+    test.each([
+        {
```
Comment (waabid) [inline src/tests/unit/tests/DetailsView/components/settings-panel/settings-panel.test.tsx]: Any way we can make these tests a bit more readable?

File: src/DetailsView/components/settings-panel/settings-panel.tsx (hunk 1 above)
Comment (smoralesd) [reply src/DetailsView/components/settings-panel/settings-panel.tsx]: I get your point, but there is a settings folder inside settings-panel... I would prefer to rename the class. Panel sounds too generic though. Any other ideas?

File: src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx (hunk 2 above)
Comment (smoralesd) [reply src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx]: They are not... I was playing with typescript version the other day and 3.x complains about this... not sure why 2.x does not

File: src/tests/unit/tests/DetailsView/components/settings-panel/settings-panel.test.tsx (hunk 3 above)
Comment (smoralesd) [reply src/tests/unit/tests/DetailsView/components/settings-panel/settings-panel.test.tsx]: how so? The properties of the class are pretty much what you need to pass as properties down the specific settings components.

Comment (smoralesd) [reply src/tests/unit/tests/DetailsView/components/settings-panel/settings-panel.test.tsx]: The test case type is gone now. Only moving part now is `isPanelOpen`

File: src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import { TextField } from 'office-ui-fabric-react/lib/TextField';
+import * as React from 'react';
+import { FlaggedComponent } from '../../../../../common/components/flagged-component';
+import { FeatureFlags } from '../../../../../common/feature-flags';
+import { NamedSFC } from '../../../../../common/react/named-sfc';
+import { SettingsProps } from '../settings-props';
+
+export const BugFilingSettings = NamedSFC<SettingsProps>('BugFilingSettings', props => {
+    const getBugSettingsUx = () => {
+        return (
```
Comment (Shobhit1) [inline src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx]: nit: do we need a `return` here? We can use `() => ()` format to eliminate `return` here. Not a blocker though. Just my thought

File: src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    };
+
+    const getGitHubBugSettingsUx = () => {
+        return (
+            <TextField
+                className="issue-setting"
+                label="Enter desired GitHub repo link:"
+                onChange={onGitHubRepositoryChange}
+                value={getBugServiceProperty('gitHub', 'repository')}
+                placeholder="https://github.com/owner/repo/issues"
```
Comment (Shobhit1) [inline src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx]: I thought the change was made where we dont really need to put in `issues` as the url. So, will it make sense to update `placeholder` text accordingly?

File: src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    const onGitHubRepositoryChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
+        props.deps.userConfigMessageCreator.setBugServiceProperty('gitHub', 'repository', newValue);
+    };
+
+    const getBugServiceProperty = (bugService: string, propertyName: string) => {
+        const bugServicePropertiesMap = props.userConfigurationStoreState.bugServicePropertiesMap;
+        if (!bugServicePropertiesMap || !bugServicePropertiesMap[bugService]) {
+            return undefined;
+        }
+        return props.userConfigurationStoreState.bugServicePropertiesMap[bugService][propertyName];
```
Comment (Shobhit1) [inline src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx]: nit:
```
        const { bugServicePropertiesMap } = props.userConfigurationStoreState;
        if (!bugServicePropertiesMap || !bugServicePropertiesMap[bugService]) {
            return undefined;
        }
        return bugServicePropertiesMap[bugService][propertyName];
```

File: src/DetailsView/components/settings-panel/settings/settings-provider-impl.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { BugFilingSettings } from './bug-filing/bug-filing-settings';
+import { HighContrastSettings } from './high-contrast/high-contrast-settings';
+import { create } from './settings-provider';
+import { TelemetrySettings } from './telemetry/telemetry-settings';
+
+export const SettingsProviderImpl = create([TelemetrySettings, HighContrastSettings, BugFilingSettings]);
```
Comment (Shobhit1) [inline src/DetailsView/components/settings-panel/settings/settings-provider-impl.ts]: Do you think we can rename `create` to maybe tell what we are creating? Personally, `create` looks a little too general for me. Let me know what you think

File: src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx (hunk 5 above)
Comment (haonliu) [reply src/DetailsView/components/settings-panel/settings/bug-filing/bug-filing-settings.tsx]: talked offline. resolve this.

**Pull request merged**